File: ruby/trinary/trinary.rb
class Trinary
  def initialize(num)
    @num = num
  end

  def to_decimal
    if invalid?
      return 0
    end

    index = 0
    trinary = @num.split("").reverse.map {|num|
      num = num.to_i*(3**index)
      index += 1
      num
    }.reduce(:+)
  end

  private
  def invalid?
    @num.match(/[^012]/) != nil
  end
end

module BookKeeping
    VERSION = 1
end

File: javascript/trinary/trinary.js
var Trinary  = function(bits){
  this.bits = bits;
}

Trinary.prototype.toDecimal = function(){
  var base = 3;
  var binaryArr = this.bits.split("").reverse();
  var length = binaryArr.length;

  for(var i = 0; i < length; i++){
    if(parseInt(binaryArr[i]) >= 0 && parseInt(binaryArr[i]) <= 3){
      binaryArr[i] *= Math.pow(base, i);
    } else {
      return 0;
    }
  }

  return binaryArr.reduce(function(a,b){
    return a + b;
  }, 0);
}

module.exports = Trinary;
